lib/src/main/kotlin/Node.kt

```diff
@@ -1,3 +1,17 @@
-internal open class Node<TChild>(parent: NodeBase?) : NodeBase(parent) where TChild : NodeBase {
-    var children = mutableListOf<TChild>()
+abstract class Node<TAction, TSelf>(
+    val parent: TSelf?,
+    val inducingAction: TAction?
+) where TSelf : Node<TAction, TSelf> {
+    val depth: Int = if (parent == null) {
+        0
+    }
+    else {
+        parent.depth + 1
+    }
+    var n = 0
+    var reward = 0.0
+    var maxReward = 0.0
+
+    abstract fun addChild(child: TSelf)
+    abstract fun getChildren(action: TAction? = null) : Collection<TSelf>
```

Comment:
When would we ever input a TAction of null? Should it not, that to get all potential children, we must tell the node what action we are taking?

Reply:
Able to accept null argument (as default) - where it will return ALL children for all actions.